src/override/malloc.cc

```diff
@@ -221,16 +297,267 @@ extern "C"
+    if (SNMALLOC_LIKELY(p != nullptr))
+    {
+      sz = bits::min(alloc_size, sz);
+      // Guard memcpy as GCC is assuming not nullptr for ptr after the memcpy
+      // otherwise.
+      if (sz != 0)
+      {
+        memcpy(p, *ptr, sz);
+      }
+      a.dealloc(*ptr);
+      *ptr = p;
+      if (rsize != nullptr)
+      {
+        *rsize = size;
+      }
+      return allocm_success;
+    }
+    return allocm_err_oom;
+  }
+
+  /**
+   * Jemalloc 3 experimental API.  Sets `*rsize` to the size of the allocation
+   * at `*ptr`.  The third argument contains some flags relating to arenas that
+   * we ignore.
+   */
+  int SNMALLOC_NAME_MANGLE(sallocm)(const void* ptr, size_t* rsize, int)
+  {
+    auto& a = ThreadAlloc::get();
+    *rsize = a.alloc_size(ptr);
+    return allocm_success;
+  }
+
+  /**
+   * Jemalloc 3 experimental API.  Deallocates the allocation
+   * at `*ptr`.  The second argument contains some flags relating to arenas that
+   * we ignore.
+   */
+  int SNMALLOC_NAME_MANGLE(dallocm)(void* ptr, int)
+  {
+    ThreadAlloc::get().dealloc(ptr);
+    return allocm_success;
+  }
+
+  /**
+   * Jemalloc 3 experimental API.  Returns in `*rsize` the size of the
+   * allocation that would be returned if `size` and `flags` are passed to
+   * `allocm`.
+   */
+  int SNMALLOC_NAME_MANGLE(nallocm)(size_t* rsize, size_t size, int flags)
+  {
+    *rsize = round_size(JEMallocFlags(flags).aligned_size(size));
+    return allocm_success;
+  }
+
+  /**
+   * Jemalloc function that provides control over alignment and zeroing
+   * behaviour via the `flags` argument.  This argument also includes control
+   * over the thread cache and arena to use.  These don't translate directly to
+   * snmalloc and so are ignored.
+   */
+  SNMALLOC_EXPORT void* SNMALLOC_NAME_MANGLE(mallocx)(size_t size, int flags)
+  {
+    auto f = JEMallocFlags(flags);
+    size = f.aligned_size(size);
+    if (f.should_zero())
+    {
+      return ThreadAlloc::get().alloc<ZeroMem::YesZero>(size);
+    }
+    return ThreadAlloc::get().alloc(size);
+  }
+
+  /**
+   * Jemalloc non-standard function that is similar to `realloc`.  This can
+   * request zeroed memory for any newly allocated memory, though only if the
+   * object grows (which, for snmalloc, means if it's copied).  The flags
+   * controlling the thread cache and arena are ignored.
+   */
+  SNMALLOC_EXPORT void*
+    SNMALLOC_NAME_MANGLE(rallocx)(void* ptr, size_t size, int flags)
+  {
+    auto f = JEMallocFlags(flags);
+    size = f.aligned_size(size);
+
+    auto& a = ThreadAlloc::get();
+    size_t sz = a.alloc_size(ptr);
+    // Keep the current allocation if the given size is in the same sizeclass.
+    if (sz == round_size(size))
+    {
+      return ptr;
+    }
+
+    if (size == (size_t)-1)
+    {
+      return nullptr;
+    }
+
+    void* p = f.should_zero() ? a.alloc<YesZero>(size) : a.alloc(size);
+    if (SNMALLOC_LIKELY(p != nullptr))
+    {
+      sz = bits::min(size, sz);
```

Comment:
Here you are using `size` which has been expanded by `aligned_size`, but not `round_size`.  This seems strange, you probably want the original size.